### PR TITLE
add option to purge existing custom data fields from users when saving

### PR DESCRIPTION
**Pull request by czue**
was going to implement this as a management command but then seemed easy enough to do in the UI.

image

@dimagi/product / @sheelio do you see any issues with adding this? i can obviously flag it if needed but seems helpful to me.

I think @esoergel has the most context on this feature? cc @proteusvacuum @kaapstorm buddies. review how you like. :fish: / :classical_building:

### CODE REVIEW

File: corehq/apps/custom_data_fields/templates/custom_data_fields/custom_data_fields.html [hunk 1]
```diff
@@ -183,10 +183,27 @@
...
+        <fieldset>
+            <div class="col-sm-10 col-sm-offset-2">
+                <div class="checkbox">
+                    <label>
+                        <input name="remove-existing" type="checkbox" data-bind="checked: purge_existing">{% trans "Remove unused fields from existing data."%}
+                        <span class="hq-help-template"
+                              data-title="{% trans "Remove Unused Fields" %}"
+                              data-content="{% blocktrans with entity_string=view.entity_string|lower %}
+                                  If checked, this will remove any extra information from any existing {{ entity_string }}
+                                  that has data fields that are no longer used.
```
Comment (czue) [inline corehq/apps/custom_data_fields/templates/custom_data_fields/custom_data_fields.html]: i struggled a bit with this text and being too lazy to add a pluralized version of `{{ entity_string }}` to the template. open to any improvements.

**Comment by sheelio** [comment]
@czue  Seems ok.   Would update the text a bit:
"Remove any other fields from existing users.   This could break existing applications if they rely on these other fields"

**Comment by biyeun** [comment]
Looks good 👍

File: corehq/apps/custom_data_fields/edit_model.py [hunk 2]
```diff
@@ -19,6 +19,7 @@ class CustomDataFieldsForm(forms.Form):
     The main form for editing a custom data definition
     """
     data_fields = forms.CharField(widget=forms.HiddenInput)
+    purge_existing = forms.BooleanField(widget=forms.HiddenInput, required=False, initial=False)
```
Comment (esoergel) [inline corehq/apps/custom_data_fields/edit_model.py]: Oh man this code is all way too abstract.  I apologize on behalf of Ethan of 2 1/2 years ago.

File: corehq/apps/users/tasks.py [hunk 3]
```diff
@@ -172,3 +171,9 @@ def reset_demo_user_restore_task(couch_user, domain):
 
     DownloadBase.set_progress(reset_demo_user_restore_task, 100, 100)
     return {'messages': results}
+
+
+@task
+def remove_unused_custom_fields_from_users_task(domain):
+    from corehq.apps.users.custom_data import remove_unused_custom_fields_from_users
+    remove_unused_custom_fields_from_users(domain)
```
Comment (esoergel) [inline corehq/apps/users/tasks.py]: :+1:

**Review by esoergel — APPROVED** [review]
This looks good to me.  It would be nice if we could display specifically what fields (if any) would be deleted, but I imagine there'd be performance concerns - not sure if you can get that without pulling all users and doing it in python.

**Review by kaapstorm — APPROVED** [review]

**Comment by czue** [comment]
@sheelio updated text. just waiting for a build to run.... :snail:

File: corehq/apps/custom_data_fields/edit_model.py (hunk 2 above)
Comment (czue) [inline corehq/apps/custom_data_fields/edit_model.py]: hahaha it did take me a long time to figure out how some pieces worked, though the ability to be generically dropped in in multiple places is nice. i did at first try to reconcile the logic that handles the removal in the edit entity form with this, but that was too coupled to the view / request to be able to do it.

**Pull request merged**